runtime/src/iree/hal/drivers/cuda/nccl_channel.c

```diff
@@ -407,6 +407,45 @@ static iree_status_t iree_hal_cuda_nccl_submit_batch_entry(
                            "ncclRecv");
       break;
     }
+    case IREE_HAL_COLLECTIVE_KIND_PERMUTE: {
+      CUdeviceptr sendbuff =
+          iree_hal_cuda_buffer_device_pointer(
+              iree_hal_buffer_allocated_buffer(entry->send_binding.buffer)) +
+          iree_hal_buffer_byte_offset(entry->send_binding.buffer) +
+          entry->send_binding.offset;
+      CUdeviceptr recvbuff =
+          iree_hal_cuda_buffer_device_pointer(
+              iree_hal_buffer_allocated_buffer(entry->recv_binding.buffer)) +
+          iree_hal_buffer_byte_offset(entry->recv_binding.buffer) +
+          entry->recv_binding.offset;
+      int16_t sendid;
+      int16_t recvid;
+      memcpy(&sendid, &entry->param, 2);
+      memcpy(&recvid, (char*)&entry->param + 2, 2);
+      printf("param=%d sendid=%hd recvid=%hd\n", entry->param, sendid, recvid);
+      if (sendid != -1) {
+        NCCL_RETURN_IF_ERROR(
+            syms,
+            ncclSend((const void*)sendbuff, entry->element_count, datatype,
+                     sendid, comm, stream),
+            "ncclSend");
+      }
+      if (recvid != -1) {
+        NCCL_RETURN_IF_ERROR(syms,
+                             ncclRecv((void*)recvbuff, entry->element_count,
+                                      datatype, recvid, comm, stream),
+                             "ncclRecv");
+      } else {
+        // Zero out recvbuff if this rank is not receiving any data.
+        iree_device_size_t num_bytes =
+            entry->element_count *
+            iree_hal_collective_element_byte_count(entry->op.element_type);
+        CUDA_RETURN_IF_ERROR(syms,
+                             cuMemsetD8Async(recvbuff, 0, num_bytes, stream),
```

Comment:
I'm wondering if this is ok to do in the runtime or should this zeroing be done on the compiler side with a `select()` or something.

Reply:
I think this is fine here as it lets us fully model the collective operation behavior without splitting it across the boundary. You've got that documented in the operation comment ("result all zeros") and that seems safer than saying "undefined contents" so 👍 IMO (good question though! :)

Reply:
Thanks @benvanik for the review! Got it, that makes sense.